Fix for "Locked-mode restore fails with sha512 errors & installs new packages when a project's RuntimeIdentifier is changed"

cristinamanum opened the pull request:
## Bug
[Locked-mode restore fails with sha512 errors & installs new packages when a project's RuntimeIdentifier is changed](https://github.com/NuGet/Home/issues/8260)

Fixes: https://github.com/NuGet/Home/issues/8260
Regression: No  
* Last working version:  N/A 
* How are we preventing it in future: 
More unit tests.  

## Fix

Added verification to detect runtime changes. 

## Testing/Validation

Tests Added: Yes

Validation:  
Unit test and manual validation.

cristinamanum committed:
Fix for issue 8260

cristinamanum committed:
Fix for issue 8260

zivkan committed:
Merge packages.config content hash enforcement (#2923)

nkolev92 committed:
Fix the deadlock that can happen during the initialization of the VSSettings (#2919)

xavierdecoster committed:
Removed expander on DeprecationControl (#2921)

cristinamanum committed:
Unit test changes

cristinamanum committed:
Unit test changes

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -6,12 +6,13 @@
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Runtime.CompilerServices;
```
No classes from this namespace are used

cristinamanum replied:
Removed

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -6,12 +6,13 @@
 
 using FluentAssertions;
-
+using FluentAssertions.Collections;
```
No classes from this namespace are used

cristinamanum replied:
Removed

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -8546,6 +8547,415 @@ public async Task RestoreNetCore_ProjectProvidedRuntimeIdentifierGraphChange_Doe
 
+        [Theory]
+        [InlineData(new string[] { "win7-x86" }, new string[] { "win-x64" })]
+        [InlineData(new string[] { "win7-x86", "net46" }, new string[] { "win-x64" })]
```
`net46` isn't a runtime. It's not clear to me what this test's initial condition is supposed to be.

If this should have been `win-x64` instead, then good, otherwise you're missing a test case where it goes from 2 RIDs and one of the two RIDs are removed.

cristinamanum replied:
Fixed. It was intended to be win-x64.

dtivel committed:
Test:  suppress NuGet UI in Apex tests (#2926)

Resolve https://github.com/NuGet/Home/issues/8325.

dtivel committed:
Test:  enable 100 package signing tests (#2924)

Resolve https://github.com/NuGet/Home/issues/8326.

kzu committed:
Add .proj to the list of known extensions for restore (#2870)

Since it's pretty trivial to get those projects to load in VS, offering a far 
better editing experience, make it possible to do restore on them by 
default just like for `.msbuildproj`.

See https://github.com/microsoft/MSBuildSdks/issues/97

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs:
```diff
@@ -72,6 +92,15 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
 
+            // Validate the runtimes for the current project did not change.
+            var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
+            var lockFileRuntimes = nuGetLockFile.Targets.Select(t => t.RuntimeIdentifier).Where(r => r != null);
```
Do we need a distinct here? 

How does this behave when you have a multi targeting project.?

cristinamanum replied:
Good catch! It should be distinct.

zivkan commented on test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs:
```diff
@@ -276,6 +275,14 @@ public void AddPackageToAllFrameworks(params SimpleTestPackageContext[] packages
+        {
+            foreach (var framework in Frameworks)
+            {
+                framework.PackageReferences = new List<SimpleTestPackageContext>();
```
```suggestion
                framework.PackageReferences.Clear();
```

This minimises memory allocation

cristinamanum replied:
Good suggestion! Updated.

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -8546,6 +8547,415 @@ public async Task RestoreNetCore_ProjectProvidedRuntimeIdentifierGraphChange_Doe
+                r.Success.Should().BeTrue();
+            }
+        }
         private static byte[] GetTestUtilityResource(string name)
```
nitpick: empty line between methods. Also there were two places above where there were 2 consecutive empty lines.

cristinamanum replied:
I cleaned empty spaces.

zivkan reviewed (approved):
I'd prefer if the PR title described the fix. When looking at the list of all open PRs, I have no idea what issue 8260 refers to and in 6 months if I'm looking back at the commit history, I'll have no idea what that issue number refers to.

dtivel committed:
Test:  clean up after test execution (#2925)

Resolve https://github.com/NuGet/Home/issues/8328.

dtivel committed:
Test:  update package download URL (#2930)

Fix https://github.com/NuGet/Home/issues/8336.

cristinamanum committed:
Feedback

cristinamanum committed:
Feedback

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs:
```diff
@@ -46,12 +46,32 @@ public static string GetNuGetLockFilePath(PackageSpec project)
+            var lockFileFrameworks = nuGetLockFile.Targets
+                .Where(t => t.TargetFramework != null)
+                .Select(t => t.TargetFramework)
+                .Distinct();
```
I would remove the distinct here. 
I think it'd only hide potential bugs. Those values should be unique.

cristinamanum replied:
A project set as:  <PropertyGroup>
    <TargetFrameworks>netcoreapp2.2</TargetFrameworks>
    <RuntimeIdentifiers>win-x86;osx-x64</RuntimeIdentifiers>
  </PropertyGroup>

Will get three dependencies in the lock file all off them with the same framework. 

Resolving the conversation here. Please add more comments at : https://github.com/NuGet/NuGet.Client/pull/2933/files#r304109104

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -8546,6 +8550,418 @@ public async Task RestoreNetCore_ProjectProvidedRuntimeIdentifierGraphChange_Doe
+                   new NuGetFramework[]{ NuGetFramework.Parse(tfm) });
+
+                projectA.Properties.Add("RestorePackagesWithLockFile", "true");
+                projectA.Properties.Add("RestoreLockedMode", "true");
```
I'd normally expect RestoreLockedMode true and a lock file not existing to be an error. 

I feel like it'd lead the customer to a false sense of security. 

@anangaur What do you think? 

Should we file a bug for this?

anangaur replied:
I agree. If RestoreLockedMode is set and lock file doesn’t exist, it should error stating “cannot file lock file...”

cristinamanum replied:
If the lock file does not exist a lock file will be created. Why this is not as expected?

nkolev92 replied:
Because the customer is indicated they want to be in locked mode. 

We can't deliver on the promise of locked mode if there's no lock file.

cristinamanum replied:
Added a new issue: https://github.com/NuGet/Home/issues/8338

nkolev92 reviewed (commented):
Nice test coverage. 

@anangaur There's a question for you in this thread that affects the tests Cristina wrote.

dtivel committed:
Test:  clean up after test execution (#2932)

Fix https://github.com/NuGet/Home/issues/8337.

cristinamanum committed:
Small change to trigger a new test run.

cristinamanum committed:
Small change to trigger a new test run.

cristinamanum committed:
Merge branch 'dev-cmanu-8260' of https://github.com/NuGet/NuGet.Client into dev-cmanu-8260

cristinamanum commented:
Need to close this pull request due to extensive file count showed as updated after rebase. A new one will be opened for the same changes.

nkolev92 commented:
There's issues sometimes after a rebase where retargetting fixes the diff. 

Looking at your branch (my local copy), it seems like you did a merge rather than a rebase though.
